### ISSUE TITLE
Software Engineering Interviews

### ISSUE THREAD

**Issue by darwinrlo** [darwinrlo/public#2]
It is April 23rd. My interview at Google is May 19th, which is on a Tuesday. This means I have 24 days left to prepare -- just about three weeks.

I'd like to spend the final week on the distributed systems course and approximate methods for solving NP-complete problems. For my ambitions, there is a real time crunch here.

The two days before that date will be spent reviewing already-processed material. No new material will be processed during those two days. Effectively, there is a deadline of midnight the Sunday before for the processing of new material.

On the morning of the interview, I will undergo a recall & reconstruction sequence to load the most high-impact knowledge into my working memory. I will tinker with this sequence between now and then. I need a first draft of this sequence by the end of this weekend.

**Comment by darwinrlo** [comment]
Cracking the Coding Interview, p. 67:

> You have a chunk of work that's done repeatedly.

I think typically this will be optimizing or removing inner loops.

**Comment by darwinrlo** [comment]
Cracking the Coding Interview, p. 71:

> Optimize & Solve Technique #3: Simplify and Generalize

IMO, it should be "simplify and extend." And a better illustration of the technique than the one presented in the book is `4.9 BST Sequences` -- though admittedly it is a more complicated example.

**Comment by darwinrlo** [comment]
To save time, check your proposal or pseudocode against the solution in the book before coding it up before coding it up.

**Comment by darwinrlo** [comment]
Most people need to work on coding on a whiteboard. Few people seem to need to work on talking out loud as they think. My difficulties are flipped.

**Comment by darwinrlo** [comment]
I've been practicing in Java. If I could go back, I would use Python from the start. Python more closely matches my thinking. Luckily, it's not too late to switch.

OTOH, I do find programming in Java to be quite enjoyable. Scala could be a good middle ground.

I think there is a substantial increase in the odds of success with Python.

**Comment by darwinrlo** [comment]
In their implementation of binary search, the middle element is computed by (a + b)/2. As we know, a+b can overflow. There is a way to write this to reduce the incidence of overflows.

**Comment by darwinrlo** [comment]
What I really like about CtCI is that there is very little waste: The background materials are minimal, and every problem teaches me something new.

**Comment by darwinrlo** [comment]
[Fahim ul Haq](https://www.linkedin.com/in/fahimulhaq/)

The CEO of Educative was a software engineer at Facebook.

**Comment by darwinrlo** [comment]
[M-W: interpose](https://www.merriam-webster.com/dictionary/interpose)

> to place in an intervening position

**Comment by darwinrlo** [comment]
Space complexity is interesting. With time complexity, every unit of work is contributes to the total running time. Space OTOH can be freed. A lot of things might take up space, but the space complexity is the maximum amount of space used at any given time.

**Comment by darwinrlo** [comment]
I don't think it's a requirement to come up with the expected solution. In fact, it is likely better if it's different while having the same or better time and space complexities.

**Comment by darwinrlo** [comment]
Peter Thiel called out Google for hoarding cash and not knowing what to do with it. One possible answer is funding and supporting startups -- if they don't know what to do with it, give it to others who do.

Also, Google can lead the way for remote working. If you can pass the technical interviews and you're Googley, then you can work at Google no matter where you are -- even Kansas.

(Look for the video clip. Get the time and place.)

**Comment by darwinrlo** [comment]
It'd be really great to work with John Carmack at Facebook. Or on Reason. Or React. Or Yann LeCun.

Or with Peter Norvig at Google. Or Jeff Dean and Sanjay Ghemawat.

* [Flutter](https://flutter.dev/)

**Comment by darwinrlo** [comment]
Make sure you enjoy the technology you're using. For me, Coda works. But I can see how, with someone else, pen and paper would work better.

**Comment by darwinrlo** [comment]
Workflowy might be better for quiz-and-recall.

**Comment by darwinrlo** [comment]
These problems require an insight, or even a trick, in order to solve. And as much as people don't like to admit it, it took someone a long time to acquire these insights. These are plausible interview problems because, when these problems were first posed, the interviewer expected to have to give you hints in order for you to make progress.

The landscape has changed. People go into these interviews knowing these insights already. They don't need as many hints, if any at all. These are the people you will be compared with.

**Comment by darwinrlo** [comment]
### Insights

* Review (using active recall) of the insights of the field of competitive programming is the single most important thing you can do to succeed at a software engineering interview (or competitive programming). It is imperative to compile these insights and review them until they are internalized. Review is a high-return activity that most people spend very little time on, preferring instead to do large quantities of problems, which is must less efficient.

* The next most important thing is practice writing the sort of code that is involved in these problems.

### How to approach a practice problem in CtCI

* Take 1 minute to extract takeaways from the problem description. Try to do it in 30 seconds.

* Spend 14 minutes drawing out examples, making observations, and coming up with insights. If possible, describe a high-level approach. The description of the approach should truly be high-level: No code, not even pseudocode, is written; just notes, perhaps in the form of a bulleted list, will suffice.

* Spend 45 minutes going through the solution. As you read, extract takeaways and generate corresponding review questions on the fly. Try to figure out the 10% that will get 90% of the result.

* Move onto the next problem.

* Every week -- perhaps even every evening -- try your hand at a coding contest.
  * LeetCode has a contest every Saturday evening. After each contest, reflect on why things didn't go as well as you would've liked and make a plan to improve upon these things.
  * LeetCode's past contests are available online. You can set up your own contests each evening.

**Comment by darwinrlo** [comment]
Putting things into our own words is something we should all practice, starting from a young age. And it does take practice to get good at.

I remember that Jerry Cain seemed to be good at this -- he usually had a unique way of phrasing things. So did several people I knew while growing up.

Also, being able to articulate things key, for me at least, to understanding things in discrete math. (I wonder if Hungarian is better suited to express ideas in discrete math. The way to test this would be to articulate ideas from discrete math in English and get someone to translate them into Hungarian.)

Being able to say things in your own words is mentioned in passing or as afterthought. It should be something we focus on front-and-center.

**Comment by darwinrlo** [comment]
Jeff Bezos:

> I set my first meeting for 10 a.m. I like to do my high IQ meetings before lunch, like anything that’s going to be really mentally challenging, that’s a 10:00 a.m. meeting

Christoph Randler:

> People whose performance peaks in the morning are better positioned for career success, because they’re more proactive than people who are at their best in the evening.

Source: [CNBC]

**Comment by darwinrlo** [comment]
Jeff Bezos to Thrive Global:

> If you shortchange your sleep, you might get a couple of extra ‘productive’ hours, but that productivity might be an illusion. When you’re talking about decisions and interactions, quality is usually more important than quantity.

Source: [CNBC]

**Comment by darwinrlo** [comment]
I'm already pretty good at putting the pieces together. What I need are the pieces.

**Comment by darwinrlo** [comment]
It is important for me to feel both closure and progress. Marking something as done is very impactful.

Feeling unresolved hinders the feeling of momentum.

**Comment by darwinrlo** [comment]
The insights you can use when you're on the spot is the stuff at the top of your head.

**Comment by darwinrlo** [comment]
Cal Newport on the Quiz-and-Recall Method for Technical Sources:

> First, narration is key. You can’t just write out the steps to an answer. You have to lecture, as if you are presenting the solution to an imaginary class, about each step and why you are doing it.

> Second, the technical discussion questions (also discussed here) help make sure you understand the underlying concepts.

[Source](https://www.calnewport.com/blog/2007/10/17/q-a-five-key-questions-about-becoming-an-efficient-student/)

Cal Newport on technical explanation questions for complicated problems:

> Given a complicated example of a certain type of problem, a good TEQ might have you provide detailed annotation on each step; explaining the logic behind each.

[Source](https://www.calnewport.com/blog/2007/09/04/monday-master-class-use-technical-explanation-questions-when-studying-for-technical-classes/)

Cal Newport on working on internalizing which techniques apply in which situations:

> In many technical courses, a big part of the challenge is figuring out which technique to apply to a given problem. A good TEQ might have you discuss the criteria for choosing from a set of different techniques for a certain type of problem.

[Source](https://www.calnewport.com/blog/2007/09/04/monday-master-class-use-technical-explanation-questions-when-studying-for-technical-classes/)

The above could be used on the 16 patterns in Grokking.

For all its virtues, CtCI will not infrequently present a solution without explicitly explaining key aspects. You have to recognize when this happens, identify a suitable TEQ, and include it in your review.

**Comment by darwinrlo** [comment]
Mega Problem Sets:

> I take this one step further by discussing how to construct Mega Problem Sets (MPS), which include, in addition to your weekly homework, selected examples from your lecture notes. If you can answer the problems in every MPS, then you are more than prepared for your upcoming test.

The problems in CtCI are a great source for the software engineering interview prep analogue of "selected examples from your lecture." The problems in LeetCode are a bit random. The problems in CtCI are extremely well chosen as the examples in lecture would be in the ideal case.

[Source](https://www.calnewport.com/blog/2007/09/04/monday-master-class-use-technical-explanation-questions-when-studying-for-technical-classes/)

**Comment by darwinrlo** [comment]
Boundary conditions are very tricky to get right. Getting them right must be worked on explicitly.

Loops
* The initial value for the iteration variable
* The loop condition. What the inequalities should be given

**Comment by darwinrlo** [comment]
Relish the times in which you find things that don't come quickly to you -- you are spending time on the concepts you most need to spend time on.

**Comment by darwinrlo** [comment]
Do your work in a way that sets you up to *love* your work -- even if it means being less efficient in the short term. So-called efficient methods disregard an important variable: you. So-called efficient methods are only effective on a short timescale. Over the long term, the system that you enjoy will be both more effective *and* more efficient.

**Comment by darwinrlo** [comment]
Lead the interviewer to Aha! moments. Epiphanies feel highly rewarding. If you can do this, this reverses the power balance, shifting it from the interviewer to you.

Recall the skit under discussion on the first day of drama class.

**Comment by darwinrlo** [comment]
Why are we still coding up recursive algorithms by hand?

**Comment by darwinrlo** [comment]
For "true" mastery, write the algorithm using pen and paper.

**Comment by darwinrlo** [comment]
It is important to do practice problems, make mistakes, and learn from them. Avoiding pitfalls and traps helps improve efficiency and thus speed.

For example, let's say you have a function `foo`. You need two parameters, one for an array and one for an index into that array. You might be tempted to call the latter parameter `i`, i.e. `foo(arr, i)`. But this will give you trouble once you're inside the function body ー you won't be able to use `i` as an iteration variable for your `for`-loops!

Inside CtCI, you will see `index` used without explanation... this is why.

Another example is accessing a grid by row and column instead of by x- and y-coordinates. See p. 361 in CtCI.

**Comment by darwinrlo** [comment]
CtCI 8.2 -- the solution backtracks from the target position. But it seems to me you could just as easily go forward from the origin.

But backtracking does cause the points to be added to the path in the right order.

**Comment by darwinrlo** [comment]
Creating a todo list is invaluable for breaking up work into chunks that feel manageable. The chunks could even be segments of text within something you have to read. Marking a todo as `[current]` commits you to it. And marking it with an `[x]` makes you feel done with, allowing you to release your emotional "lock" on it and move on.

**Comment by darwinrlo** [comment]
Turning off automatic capitalization, spelling, and link correction:

[Correct your spelling & grammar in Google Docs](https://support.google.com/docs/answer/57859?co=GENIE.Platform%3DDesktop&hl=en)

Tools - Preferences - General

**Comment by darwinrlo** [comment]
* The purpose of taking notes during class is to capture what you need to know and what you need to figure out outside of class.
* The purpose of taking notes outside of class is to put things into your own words, which is critical for achieving understanding.

**Comment by darwinrlo** [comment]
In school, during the week, notes are taken of lecture (putting things into one's own words) and review questions are generated.

Over the weekend, review using active recall is done.

**Comment by darwinrlo** [comment]
Peter Norvig's lecture notes do a great job on graph search and even dynamic programming.

**Comment by darwinrlo** [comment]
I want to teach my queue technique. Maybe I even want to develop a software system for it.

I can demonstrate it on Lustig's neuroscience paper.

This could be a significant contribution.

This is a very engaging way to learn.

**Comment by darwinrlo** [comment]
Observation: Once a question is explicitly verbalized, my subconscious is triggered to work on it. Often, it is able to surface a feeling for what the answer might look like almost right away.

But the question has to be explicitly posed or else my subconscious won't be triggered to work on it.

**Comment by darwinrlo** [comment]
Biology class
* Didn't do well on the first biology midterm
* Achieved the mean on the second biology midterm

Came all the way back from Davis. Went through slides with him together at a Starbucks. Looked at each thing and posed questions about what these things meant. This is not just a box. It's a cell membrane.

Interpreter -- never thought so hard in his life.

**Comment by darwinrlo** [comment]
After articulating something -- either through writing or speech -- I am able to see a picture in my head. It seems that the paintbrush for the canvas in my brain is explicitly articulating thoughts using language.

**Comment by darwinrlo** [comment]
I really need to work on not making off-by-one errors. I need to compile a list of representative problems that exercise nuances that lead to off-by-one errors and do them at the start of every day.

It's a good idea in general, I think, to do quiz-and-recall on important things.

**Comment by darwinrlo** [comment]
I believe this reasoning is incorrect.

> Since, in each step, the number of subsets doubles as we add each element to all the existing subsets, the time complexity of the above algorithm is O(2^N), where ‘N’ is the total number of elements in the input set.

[Educative: Subsets (easy)](https://www.educative.io/courses/grokking-the-coding-interview/gx2OqlvEnWG)

This seems to assume creating a copy of an `ArrayList` (and creating a new subset by adding another element) takes O(1) time. In fact, it is proportional to the size of the array.

If linked lists were used, then it would take O(1) time to create a new list with a new head. But this is not what was done.

**Comment by darwinrlo** [comment]
The brute-force solution doesn't actually enumerate the items. It just gives the achievable "profit."

[Educative: 0/1 Knapsack (medium)](https://www.educative.io/courses/grokking-the-coding-interview/gkZNLjV2kBk)

**Comment by darwinrlo** [comment]
The optimal substructure of a dynamic programming problem is just the recursive structure.

**Comment by darwinrlo** [comment]
I discovered something that worked -- it took a long time -- when I was taking a class in digital systems.

**Comment by darwinrlo** [comment]
Feeling on the spot -- or otherwise judged -- results in impaired ability to perform cognitive tasks. If you feel judged inappropriately, it is something you must overcome -- perhaps through exposure therapy.

**Comment by darwinrlo** [comment]
> This means that dp[i][c] will represent the maximum knapsack profit for capacity ‘c’ calculated from the first ‘i’ items.

I think it should be the "maximum knapsack profit" achievable using items i and after.

[0/1 Knapsack (medium)](https://www.educative.io/courses/grokking-the-coding-interview/gkZNLjV2kBk)

**Comment by darwinrlo** [comment]
Consider the following two ways to pose a question.

1. Why do we have to do X?
2. Do Y. (And expect that you will remember to do X.)

1 is a good initial question. Make it a goal to advance to 2.

**Comment by darwinrlo** [comment]
1. Go through the material. Put things into my own words.
2. Pull out takeaways.
3. Create quizzes on the takeaways.
4. Take the quizzes. Answer questions using speech. (This is the single most high-return activity during prep.)

**Comment by darwinrlo** [comment]
I'm already good at putting the pieces together. I just need to learn what the pieces are.

**Comment by darwinrlo** [comment]
Students can understand things by writing code.

Set theory is also important. So is abstract algebra. And perhaps topology?

**Comment by darwinrlo** [comment]
Discrete math, algorithms and data structures, and the theory of computation are weight-lifting for the brain.

Developing a terminology is very important.

**Comment by darwinrlo** [comment]
I'd like to call them "review prompts." "Prompts" because they may not be questions. And their purpose is to give you an idea of what to summon into working memory.

Recalling something elevates its priority to your brain. It also identifies the specific points where you need to get clarification.

**Comment by darwinrlo** [comment]
If I could go back, I would start with the Big-O and technical interview chapters in CtCI.

**Comment by darwinrlo** [comment]
It's important for two recall sessions to have a full night of sleep in between.

After a recall session, make sure to get a full night of sleep that night.

**Comment by darwinrlo** [comment]
If you're struggling with designing the high-level algorithm, don't worry about the edge cases. The former is hard enough without having to worry about the latter. You can get to edge cases once you're confident in your ability to design algorithms at a high-level.

**Comment by darwinrlo** [comment]
If I work at Google, I can study Jeff Dean's code. Peter Norvig is also there.

John Carmack is at Facebook.

I want to develop tools to improve the output of the entire company -- just like MapReduce did.

DeepMind is part of Google. Also, it's in London, which is really cool.

**Comment by darwinrlo** [comment]
If you don't know how to learn, learn how to learn first before earnestly preparing for interviews. Interview prep can be used to learn how to learn.

**Comment by darwinrlo** [comment]
Also important is learning how to do. For some people, learning is rewarding for its own sake. For most people -- including myself to a large extent -- learning is valuable to the extent that it enables me to do things I want to do.

The week, for many, feels natural as an execution cycle.

**Comment by darwinrlo** [comment]
[Hacking a Google Interview](http://courses.csail.mit.edu/iap/interview/index.php)

This is very condensed. Probably good to start with.

A strategy is to look for others who are more experienced to distill, condense, and filter the material for you. You should be able to exceed those who have gone before you -- not just match their performance and timeline.

**Comment by darwinrlo** [comment]
Alexa can be used to take voice memos.

**Comment by darwinrlo** [comment]
Break things up into `solid`, `near-solid`, and `shaky`.

* Material that you have just created recall prompts for should go into `shaky`.
* After two recall sessions (done on different days), the material can be moved to `near-solid`.
* If after a week has gone by since moving the material to near-solid you are able to conduct a successful recall session, then you can move the material to `solid`.
* For a review session of a given body of material to count, it has to be done on a different day than the generation of the recall prompts for that body of material and other review sessions for that body of material.
* How you break up the material is very personal to you. One natural grouping is by problem.
* Don't let let the `shaky` list get too big. When the `near-solid` list gets big, you should focus on moving material from `near-solid` to `solid`.

Reference Julia's article.

**Comment by darwinrlo** [comment]
With the proliferation of products and content to help with interview preparation, the performance of the field has improved markedly -- and consequently, expectations at the top companies, such as AirBnb, Facebook, and Google, have gotten much higher.

CtCI says: (quote about how your performance is judged relative to others' performance)

Facebook goes right out and says you need to be able to confidently do two Medium problems from Leetcode within 35 minutes.

**Comment by darwinrlo** [comment]
Schedule a mock interview using Gainlo for a month out.
Each week, prepare for the next LeetCode competition.
Start by going through Grokking the Coding Interview.

**Comment by darwinrlo** [comment]
During my interviews at Facebook, I did well at the coding interviews. But I failed the system design interview. I had never heard of them and didn't know what they were.

**Comment by darwinrlo** [comment]
[Grokking: TinyURL](https://www.educative.io/courses/grokking-the-system-design-interview/m2ygV4E81AR)

> To cache 20% of these requests

I don't think caching 20% of requests is a particularly relevant thing to do. Rather, 80% of requests will be for 20% of the short links of the day. This will take 20% of the memory that would be needed to store objects if each request was for a distinct resource.

**Comment by darwinrlo** [comment]
Without a rigid structure, motivation can lapse -- even if you have deadlines set up.

You feel like you're digging a bottomless hole. This is a bad feeling. If you feel this way, you need to change something.

You need both deadlines and structure.

At the same time, there has to be room for improvisation -- unplanned work. Space needs to be created for luck and breakthroughs to happen.

Getting all the pieces together is difficult. Even knowing what they are is difficult!

The weekend and evenings are terrific candidates for improvisation.

**Comment by darwinrlo** [comment]
Half-hour to hour-long session to set up your mindset for the day.

* Remind yourself of the why.
* Remind yourself of your plan.

**Comment by darwinrlo** [comment]
Have a separate project management system for each session.

**Comment by darwinrlo** [comment]
It takes time to understand things. You need to do what you need to do to get to understanding. You can't be anymore efficient than that.

Your understanding has to take the form of a story.

**Comment by darwinrlo** [comment]
You need to work in fixed-time blocks (and to take breaks). You need to evaluate how you did and how you need to improve on the next block. You need to lift your head from the weeds and make sure you're still pointed in the right direction.

**Comment by darwinrlo** [comment]
Grokking the Coding Interview is a huge step forward for interview preparation and competitive programming training. Someone with a lot of experience has distilled a vast body of knowledge into the tightest presentation of the materials ever created.

**Comment by darwinrlo** [comment]
Have the courage to follow half-baked ideas. "Take a leap of faith."

**Comment by darwinrlo** [comment]
It is more efficient and more effective to work on being able to write the story and then to be able to tell the story using speech than it is to go straight to speech.

**Comment by darwinrlo** [comment]
It takes several attempts on different days to write the story that makes you feel resolved, in particular, makes you feel that you understand.

The process of creating the story leads to making realizations and connections.

It takes several drafts.

Creating a diagram may be another way to achieve the same thing.

**Comment by darwinrlo** [comment]
CtCI is on the easy side. You may be able to get into Google but not necessarily at a high rank or salary.

But it's a decent place to start.

**Comment by darwinrlo** [comment]
Example of a reflection:

I would get deep in thought. Then I'd "come to" and not know where I am in the session.

* I can use a timer so I can see how much time is remaining.
* Or I can use a system involving the time, which I can see in the upper-right corner of my screen at nearly all times.

**Comment by darwinrlo** [comment]
Meditating after a work bout releases your mental resources. It makes you feel fresh and clear-headed for the next work bout. It also pulls you out of the weeds so you can see the big picture again.

**Comment by darwinrlo** [comment]
Turn reflections from one work bout to intentions (or even commitments) in the next work bout. Iterate quickly. Don't have to wait a whole day to make changes.

**Comment by darwinrlo** [comment]
You have to finalize the current version before you can iterate on it.

Instead of working on the next version, you could be working on the current version well past the point of diminishing returns.

It can be anything that gives you the feeling of closure.
* Sending an email to yourself
* Showing someone
* Posting to Facebook
* Hitting save on a wiki entry to conclude the current set of edits

Closure: releasing mental resources and freeing them up for other uses

**Comment by darwinrlo** [comment]
I don't really need Python. I have Python-like pseudocode already.

**Comment by darwinrlo** [comment]
[Monday Master Class: Conquer Complicated Material with the Mini-Textbook Method](https://www.calnewport.com/blog/2008/06/23/monday-master-class-conquer-complicated-material-with-the-mini-textbook-method/)

> A reader recently asked me for some study advice. He was facing an exam in a course with unusually complicated material. The concepts were numerous, and tricky to understand, and connected to each other in non-obvious ways. It was clear that there was too much information to be efficiently handled by standard quiz-and-recall, so I referred him to my favorite under-appreciated study technique: the focused cluster method.

> This was still, however, not enough. As the reader was quick to observe, there was so much material connected in so many different ways that even creating a quick rapid-fire question for each key point would soon spiral out of control. There would be way too many review questions.

> When faced with a course with large volumes of complicated material, reduce your notes to a collection of textbook-style chapters. Write these like a real textbook. That is, use complete sentences and logical explanations.

> Your goal should be to reduce and synthesize. A good rule of thumb is to have at most one succinct chapter per each week of notes.

> Things you might include in the sample chapter:

> A high-level description of the concepts covered in the chapter.

> A list of definitions.

> Good, succinct descriptions of the big ideas, theories, or frameworks.

> A discussion of how the different elements from the previous item connect or compare and contrast.

> The chapter writing process itself provides a powerful review, as it helps you construct a structure that transforms copious notes into coherent and compact form that is easier to review.

> The next step of the process is to construct a chapter prompt sheet for each of these chapters. On the prompt sheet, record a basic outline for the chapter.

> Finally, to review, do the following. For each chapter consider the corresponding outline. Load up your favorite word processor, and, using only the outline as a guide, attempt to type, from scratch, a new draft of the textbook chapter. Don’t peek at the original chapter.

> After your done, check your result against the original chapter; just like in the quiz-and-recall method, go back and try again later if there are areas where you had trouble.

> For classes with a large volume of complicated, interconnected material, the advantages of this method are two-fold. First, condensing the material into textbook chapters reduces the amount of information to review. A synthesized chapter will be more succinct than a long multi-page list of the type of rapid-fire questions used in a technique like the focused-cluster method.

> Second, typing the sample textbook chapter can prove quicker than trying to explain things out loud.The reason: it’s easier to express complicated ideas by typing rather than speaking. With typing, you can edit sentences, and go back and rearrange your structure as needed. When speaking, on the other hand, if the concepts are tricky and connected in intricate ways, you’re prone to getting tripped up.

> This technique might be overkill in many situations. For upper-level classes, however, writing your own textbook from scratch, though somewhat slow, might still be the fastest way to actually master the material.

**Comment by darwinrlo** [comment]
Give examples of applying Cal Newport's techniques to interview prep, e.g. focused question clusters or even the mini-textbook method for system design concepts.

Give focused question clusters a shot before attempting the mini-textbook method.

**Comment by darwinrlo** [comment]
[Monday Master Class: Use Focused Question Clusters to Study for Multiple Choice Tests](https://www.calnewport.com/blog/2007/09/24/monday-master-class-use-focused-question-clusters-to-study-for-multiple-choice-tests/)

> From experience, I know that it can take a long time to transform your notes into the focused question clusters. This follows directly from the volume of rapid-fire questions you end up having to record. To keep things painless, it’s highly recommended that you consider transforming your notes into these clusters every week as you proceed through the term. This will keeps the studying itself a reasonable chore.

**Comment by darwinrlo** [comment]
[You Know What You Write: The Textbook Method for Ultra-Learning](https://www.calnewport.com/blog/2012/08/10/you-know-what-you-write-the-textbook-method-for-ultra-learning/)

The comparison highlights the power of a simple act: describing and organizing information in your own words.

**Comment by darwinrlo** [comment]
[Building Secure & Reliable Systems](https://landing.google.com/sre/books/)

**Comment by darwinrlo** [comment]
I think annotating code lists could be the way to go. Maybe quiz-and-recall questions for very important takeaways.

**Comment by darwinrlo** [comment]
Ben Lynn:

> I’m fortunate I’ve always intuitively known how to attempt problems that I have no idea how to start. Well, “no idea” is inaccurate: I usually have some ideas, it’s just that they seem unlikely to work. But I’ll pursue them anyway, and with luck, understand more about the problem. With even more luck, this leads to more ideas, which eventually crack open the problem.

[Ben Lynn: How I Solve It](https://crypto.stanford.edu/~blynn/haskell/solve.html)

**Comment by darwinrlo** [comment]
This might be the way to go:

> Annotating a complicated example. Given a complicated example of a certain type of problem, a good TEQ might have you provide detailed annotation on each step; explaining the logic behind each.

Or just the key takeaways.

Source: [Monday Master Class: Use Technical Explanation Questions When Studying For Technical Classes](https://www.calnewport.com/blog/2007/09/04/monday-master-class-use-technical-explanation-questions-when-studying-for-technical-classes/)

**Comment by darwinrlo** [comment]
The functional equivalent of underlining an online text might be clipping out quotes and screenshots.

**Comment by darwinrlo** [comment]
A common path seems to be starting by solving the problem by assuming there are no duplicates and then evolving the solution to handle duplicates.

**Comment by darwinrlo** [comment]
# Weekends

[pgbovine: Working on Weekends](http://pgbovine.net/working-on-weekends.htm):

> I then came up with a better policy: I do only work that I want to do on weekends, not work that I need to do. In other words, I pick only tasks that I'm genuinely interested in doing, not those that I need to do out of professional obligation. As a result, sometimes I work a ton on weekends, and sometimes I barely work at all. But no matter what, I'm always doing what I want to do, not what I need to do.

> As a result of these two rules, I often get my best work done on weekends because, unlike on weekdays, I have complete freedom to work on what I want without getting interrupted by tasks that other people put on my queue. At the same time, there is never any pressure to make real progress since, after all, it's my weekend! Any progress that I do make is a bonus.

> The lack of immediate pressure enables me to try more unconventional approaches to my research on weekends. If whatever I attempt doesn't pan out, then that doesn't matter, because I was just playing around anyhow. If I'm in the mood, I skim through papers, browse technical websites, test new open-source software projects, draw concept sketches in my notebook, and do other vaguely research-related things for fun. The important thing is that my weekend work doesn't feel like work. I'm convinced that this sort of “playful quasi-work” is essential for making interesting creative leaps rather than just predictable incremental steps.

[Rhadika Nagpal: How I Learned to Stop Worrying and Love the Tenure-Track Faculty Life](https://blogs.scientificamerican.com/guest-blog/the-awesomest-7-year-postdoc-or-how-i-learned-to-stop-worrying-and-love-the-tenure-track-faculty-life/)

> And at that point I decided that 50 would just have to be enough.

> But of course less hours mean you get to do less work.

> So eventually I came up with an easier solution. I decided on a priori "fixed amounts" in which I was allowed to agree to do things. Once the quota is up, I have to mandatorily say no.

> I stop working late Friday night and I don't open my email client until Monday morning.

**Comment by darwinrlo** [comment]
The pseudocode used in many algorithms explanations is too far removed from actual code to help with assessing time complexity. This is something I think should change about how algorithms is taught.

**Comment by darwinrlo** [comment]
If you're interested in a topic, then the time-to-understanding and the retention are both boosted. It's worth factoring interest into the priority of a study topic.

Interest is a critical factor in "genius."

**Comment by darwinrlo** [comment]
Don't be focused on outcomes. Focus on the value intrinsic to the experience.

**Comment by darwinrlo** [comment]
Learn in Java. Then translate to Python.

**Comment by darwinrlo** [comment]
Whenever you have emotional difficulty making the right decision, say to yourself:

#gooddecisionmaking

Or: #chooseoutcomesnotactions

**Comment by darwinrlo** [comment]
This is absolutely terrific for coding up solutions.

https://repl.it/

You can link to code snippets from Workflowy.

**Comment by darwinrlo** [comment]
If something strikes your interest, your ability to learn is significantly increased. This should factor into your prioritization as well.

**Comment by darwinrlo** [comment]
This does not make sense. It is likely wrong.

> At every step of the algorithm, we find a vertex which is in the other set (set of not yet included) and has a minimum distance from the source.

[Geeks for Geeks: Dijkstra's shortest path algorithm](https://www.geeksforgeeks.org/dijkstras-shortest-path-algorithm-greedy-algo-7/)

> Assign distance value as 0 for the source vertex so that it is picked first.

I don't like this. Just initialize the shortest path tree with the source vertex.

**Comment by darwinrlo** [comment]
The heap interface in Python really sucks. I guess because it's a heap interface and not a priority queue interface.

**Comment by darwinrlo** [comment]
Lectures are very helpful to professors. It is a forcing function for them to decide which concepts are most useful and to work on explaining them simply, which forces them to level up in terms of their understanding.

Also, a class is a precursor for a book.

One-off talks are precursors to classes.

**Comment by darwinrlo** [comment]
The pseudocode for a heap-based implementation of Dijkstra's is given in Algorithms Illuminated II, p. 109. I believe it to be incorrect.

In the first line of the first loop iteration, the minimum element is extracted from the heap. At this point, all vertices besides the starting point have a key of INF and are inside the heap. I assume that one is arbitrarily selected and returned.

I was incorrect. The key of the origin vertex is 0, which means it is the element that is extracted from the heap in the first iteration of the "main" loop.

**Comment by darwinrlo** [comment]
Err....

Consistent Hashing, Grokking: "For load balancing, as we discussed in the beginning, the real data is essentially randomly distributed and thus may not be uniform."

**Comment by darwinrlo** [comment]
Another word for the top of one's head: working memory.

**Comment by darwinrlo** [comment]
[Grokking: Slide Window Median](https://www.educative.io/courses/grokking-the-coding-interview/3Y9jm7XRrXO)

They use a convoluted, allegedly O(log k) way to delete an element, but they search for the element first in O(k) time anyway.

**Comment by darwinrlo** [comment]
[Grokking: All Paths for a Sum](https://www.educative.io/courses/grokking-the-coding-interview/B815A0y2Ajn)

The return result in the example is a count of paths. But the code returns the actual paths.

"and for every leaf node we might have to store its path which will take O(N)."
* I'm not sure this is true. It's just adding a pointer to a list. There is no copy. It's O(1).

The space complexity discussion does not seem correct.

**Comment by darwinrlo** [comment]
### Problem-Solving Loop
* Synthesize speech from your hunches.
* Take a hunch -- prioritize ones that seem conventional -- and synthesize speech from it.
* Take the speech and synthesize pseudocode from it.
* Test the pseudocode. Come up with test cases, including boundary cases and test cases for each branch, and execute the pseudocode on the test cases. This will entail drawing diagrams.
* If the proposal works and is O(n log n) or faster, you're done!
* Otherwise:
  * If the proposal doesn't work, look at the test cases that break the pseudocode and characterize (using speech) why the proposal doesn't solve the problem.
  * If the solution is too slow, identify bottlenecks, operations that are done repeatedly inside an inner loop, and duplicate work being done.
  * If you get another hunch from following this process, take another loop around this cycle. Otherwise, start going through algorithms, data structures, and patterns to see if any of them can be applied to the problem.

**Comment by darwinrlo** [comment]
Grokking:

> Coding interviews are getting harder every day. A few years back, brushing up on key data structures and going through 50-75 coding interview questions was more than enough prep for an interview. Today, everyone has access to massive sets of coding problems, and they've gotten more difficult to account for that. The process has gotten more competitive.

[Source](https://www.educative.io/courses/grokking-the-coding-interview)

**Comment by darwinrlo** [comment]
### Week 1

Grokking
* Merge Intervals 
  * Merge Intervals (medium)
  * Insert Interval (medium)
  * Intervals Intersection (medium)
* Tree Breadth First Search
  * Binary Tree Level Order Traversal
  * Reverse Level Order Traversal
  * Zig Zag Traversal
* 0/1 Knapsack Dynamic Programming
  * Equal Subset Sum Partition (medium)
  * Subset Sum (medium)
  * Minimum Subset Sum (hard)
* Top K Elements
  * K Closest Points to the Origin
  * Connect Ropes
  * Top K Frequent Numbers

**Comment by darwinrlo** [comment]
Figuring out what to each day in the morning sucks. Choose a suitable execution cycle and plan out the whole cycle so that there is very little or no uncertainty about what you will be doing. New things have to beat what you're already doing to be slotted in.

**Comment by darwinrlo** [comment]
The patterns do really have to be strong in your mind. Repeated reconstruction is how you strengthen a pattern in your mind.

**Comment by darwinrlo** [comment]
It's almost impossible to solve a difficult problem when I need to go to the bathroom. Make sure not to drink too much coffee. And get over feeling awkward about taking a bathroom break during the interview.

**Comment by darwinrlo** [comment]
When a variable is a number, be wary of testing for falsity. 0 evaluates to False. So if you mean to check for None, check for None.

For lists, checking for falsity is a great way to handle the None and empty-list cases using one expression.

**Comment by darwinrlo** [comment]
To really understand a problem, you need to do a write-up for it. And you need to keep refining the write-up. Due to the effort and time that needs to be invested to create a write-up, you can only do this for a relatively small number of problems so you have to be selective about which problems you do this for.

Even better, repeatedly reconstruct the write-up from scratch. The write-ups will get better, which will indicate better and better understanding. You'll think of better ways to think about the problem, which will manifest as how you phrase and denote things.

**Comment by darwinrlo** [comment]
Posing a question as an interviewer helps you get solid on the problem as well.

**Comment by darwinrlo** [comment]
I need to emphasize coming up with test cases.
